addons/addon-base-raas/packages/base-raas-services/lib/data-source/access-strategy/roles-only/__tests__/filesystem-role-service.test.js

```diff
@@ -359,5 +466,170 @@ describe('DataSourceBucketService', () => {
+
+      // CHECK (Attempt #2)
+      expect(usageService.removeUsage).toHaveBeenCalledTimes(4);
+      expect(service.saveEntity).toHaveBeenCalledTimes(4);
+      expect(deleteObj.delete).toHaveBeenCalledTimes(1);
+    });
+
+    it('Role policy retry after unsuccessful update attempt - updateAssumeRolePolicy idempotency', async () => {
+      // BUILD
+      const requestContext = createAdminContext();
+      const study = createStudy();
+      const env = { id: 'env-1' };
+      const memberAcct = '1234456789012';
+      const arn = 'fs-role-arn';
+      const fsRoleEntity = {
+        arn,
+        studies: {
+          'study-1': {
+            accessType: 'readwrite',
+            envPermission: { read: true, write: true },
+            folder: '/',
+            kmsArn: undefined,
+            kmsScope: 'none',
+          },
+        },
+        trust: [memberAcct, '3333333'],
+      };
+      service.deprovisionRole = jest.fn();
+      usageService.removeUsage = jest.fn(_rq => {
+        return Promise.resolve({ items: [], removed: true });
+      });
+
+      service.saveEntity = jest.fn((_rq, entity) => {
+        return Promise.resolve(entity);
+      });
+      service.find = jest.fn().mockResolvedValue(fsRoleEntity);
+      const deleteObj = {
+        key: () => {
+          return deleteObj;
+        },
+        delete: jest.fn(),
+      };
+      service._deleter = jest.fn().mockReturnValue(deleteObj);
+
+      // EXECUTE
+
+      // Attempt #1 (Unsuccessful)
+      const error = new Error('Some error during updating IAM role policy');
+      error.boom = true;
+      error.safe = true;
+      service.updateAssumeRolePolicy = jest.fn(async () => {
+        return Promise.reject(error);
+      });
+      await expect(service.deallocateRole(requestContext, arn, study, env, memberAcct)).rejects.toThrow(error);
+
+      // CHECK (Attempt #1)
+      expect(usageService.removeUsage).toHaveBeenCalledTimes(1);
+      expect(deleteObj.delete).not.toHaveBeenCalled();
+      expect(service.saveEntity).not.toHaveBeenCalled();
+
+      // Attempt #2 (Successful)
+      service.updateAssumeRolePolicy = jest.fn();
+      await service.deallocateRole(requestContext, arn, study, env, memberAcct);
+
+      // CHECK (Attempt #2)
+      expect(usageService.removeUsage).toHaveBeenCalledTimes(2);
+      expect(deleteObj.delete).not.toHaveBeenCalled(); // There are still other accounts linked
+      expect(service.saveEntity).toHaveBeenCalledTimes(1);
+    });
+
+    it('Delete role policy with missing study and account trust', async () => {
+      // BUILD
+      const requestContext = createAdminContext();
+      const study = createStudy();
+      const env = { id: 'env-1' };
+      const memberAcct = '1234456789012';
+      const arn = 'fs-role-arn';
+      const fsRoleEntity = {
+        arn,
+        studies: {},
+        trust: [],
+      };
+      service.updateAssumeRolePolicy = jest.fn();
+      service.deprovisionRole = jest.fn();
```

Comment:
is `deprovisionRole` being unit tested anywhere?

Reply:
Doesn't look like it, I'll add it.